pennylane_lightning/lightning_kokkos/lightning_kokkos.py

```diff
@@ -13,148 +13,405 @@
+
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        state (LightningKokkosStateVector): handle to the Lightning state vector
+        batch_obs (bool): Determine whether we process observables in parallel when
+            computing the jacobian. This value is only relevant when the lightning
+            kokkos is built with OpenMP. Default is False.
+        wire_map (Optional[dict]): a map from wire labels to simulation indices
+
+    Returns:
+        TensorLike: The Jacobian of the quantum script
+    """
+    if wire_map is not None:
+        [circuit], _ = qml.map_wires(circuit, wire_map)
+    state.reset_state()
+    final_state = state.get_final_state(circuit)
+    return LightningKokkosAdjointJacobian(final_state, batch_obs=batch_obs).calculate_jacobian(
+        circuit
+    )
+
+
+def simulate_and_jacobian(
+    circuit: QuantumTape, state: LightningKokkosStateVector, batch_obs=False, wire_map=None
+):
+    """Simulate a single quantum script and compute its Jacobian.
 
-def _kokkos_dtype(dtype):
-    if dtype not in [np.complex128, np.complex64]:  # pragma: no cover
-        raise ValueError(f"Data type is not supported for state-vector computation: {dtype}")
-    return StateVectorC128 if dtype == np.complex128 else StateVectorC64
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        state (LightningKokkosStateVector): handle to the Lightning state vector
+        batch_obs (bool): Determine whether we process observables in parallel when
+            computing the jacobian. This value is only relevant when the lightning
+            kokkos is built with OpenMP. Default is False.
+        wire_map (Optional[dict]): a map from wire labels to simulation indices
+
+    Returns:
+        Tuple[TensorLike]: The results of the simulation and the calculated Jacobian
+
+    Note that this function can return measurements for non-commuting observables simultaneously.
+    """
+    if wire_map is not None:
+        [circuit], _ = qml.map_wires(circuit, wire_map)
+    res = simulate(circuit, state)
+    jac = LightningKokkosAdjointJacobian(state, batch_obs=batch_obs).calculate_jacobian(circuit)
+    return res, jac
+
+
+def vjp(
+    circuit: QuantumTape,
+    cotangents: Tuple[Number],
+    state: LightningKokkosStateVector,
+    batch_obs=False,
+    wire_map=None,
+):
+    """Compute the Vector-Jacobian Product (VJP) for a single quantum script.
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        cotangents (Tuple[Number, Tuple[Number]]): Gradient-output vector. Must
+            have shape matching the output shape of the corresponding circuit. If
+            the circuit has a single output, ``cotangents`` may be a single number,
+            not an iterable of numbers.
+        state (LightningKokkosStateVector): handle to the Lightning state vector
+        batch_obs (bool): Determine whether we process observables in parallel when
```

Comment:
Does LK support `batch_obs`?

Reply:
I am not sure but LQ has the same `batch_obs` option. I think we can leave this option to keep the same structures between Devices and, if it is possible, implement the batch option in the future. What do you think? :thinking: